### PR TITLE
Make oc debug return the cointainer's return code

### PR DESCRIPTION
**Pull request by juanluisvaladas**
Fixes https://github.com/openshift/origin/issues/15854

### CODE REVIEW

**Comment by juanluisvaladas** [comment]
Is the  automatic commit expected or did I do something wrong?

File: pkg/oc/cli/cmd/debug.go [hunk 1]
```diff
@@ -362,12 +362,23 @@ func (o *DebugOptions) Debug() error {
...
 				stderr = os.Stderr
 			}
+
+			podResult, err := o.Attach.PodClient.Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+			var rc int
+			if err != nil {
+				fmt.Fprintf(stderr, "error: unable to get the pod return code.")
+				rc = 1
+			} else {
+				rc = int(podResult.Status.ContainerStatuses[0].State.Terminated.ExitCode)
```
Comment (juanvallejo) [inline pkg/oc/cli/cmd/debug.go]: I think it might be worth checking for `len(podResult.Status.ContainerStatuses)`

Comment (juanvallejo) [reply pkg/oc/cli/cmd/debug.go]: Also, `podResult.Status.ContainerStatuses[*].State.Terminated` can be nil, causing a panic, and the debug pod not to be deleted. Worth checking for that as well. Maybe do:
```go
if err == nil && len(podResult.Status.ContainerStatuses) > 0 && podResult.Status.ContainerStatuses[0].State.Terminated != nil {
    rc = int(podResult.Status.ContainerStatuses[0].State.Terminated.ExitCode)
} else {
    fmt.Fprintf(stderr, "error: unable to get the pod return code.")
    rc = 1
}
```

**Comment by juanvallejo** [comment]
cc @fabianofranz

File: pkg/oc/cli/cmd/debug.go (hunk 1 above)
Comment (juanluisvaladas) [reply pkg/oc/cli/cmd/debug.go]: If no -c flag isn't provided only the first one will be used so **IF I understand it correclty** it's safe  be safe.

However, if the -c flag is provided then it'd be necessary to iterate over the containerStatutes and this should be blocked until implemented.

Comment (juanluisvaladas) [reply pkg/oc/cli/cmd/debug.go]: Also, is it possible under any scenario that we get a 0 length causing the program to read out of index?

**Comment by juanluisvaladas** [comment]
Also I'm not quite sure running there the os.Exit is actually a good idea.

**Comment by juanluisvaladas** [comment]
Now it will iterate for the pods over each container statuses. Works with pods with multiple containers or just one container.

**Comment by smarterclayton** [comment]
Please add an extended test for debug

**Comment by jperville** [comment]
Hello @juanluisvaladas , possible to run the retest so that this PR has a chance to be merged? Thank you very much.

**Comment by juanluisvaladas** [comment]
@jperville I'm pretty certain this was a flake  so it's pretty likely that rerunning this fixes it. 

However I have to write an extended test and that is going to take a while because:
1- I have to add a function to github.com/openshift/origin/test/extended/util
2- Add a couple tests to test/extended/cli/debug.go
3- Run the tests and make sure they do work.

The third point is the slow one. There must be I can a way to run only this test  but I haven't looked into how the extended tests work internally. I have holidays next week so Ill have a look then

**Comment by juanluisvaladas** [comment]
@jperville I'm pretty certain this error was a flake  so it's pretty likely that rerunning this fixes it. 

However I have to write an extended test which involves:

1- Add a function to github.com/openshift/origin/test/extended/util to verify the exit code of the command??
2- Add  a few tests  (4?) to test/extended/cli/debug.go
3- Run the tests and make sure they do work as expected. (This is actually pretty slow)

It will take a few hours. I have holidays in a couple weeks so I'll have a look at it then.

**Comment by juanvallejo** [comment]
@juanluisvaladas 
> The third point is the slow one. There must be I can a way to run only this test but I haven't looked into how the extended tests work internally

You can specify the `SUITE` and `FOCUS` args when running `make test-extended`:

```
$ make test-extended SUITE=conformance|networking|... FOCUS=deployments|pods|images...
```

**Comment by juanluisvaladas** [comment]
Didn't add a test yet, no new changes don't review.

**Comment by juanvallejo** [comment]
@juanluisvaladas friendly ping

**Comment by juanluisvaladas** [comment]
@juanvallejo I've been delaying this far too long, I'm getting a failure while running the tests I don't understand:

```
FAILURE after 59.592s: test/extended/cmd.sh:119: executing 'oc get pods/new-pull-pod -o jsonpath='{.status.containerStatuses[0].imageID}'' expecting any result and text 'docker'; re-trying every 0.2s until completion or 60.000s: the command timed out
Standard output from the command:
Standard error from the command:
[ERROR] PID 13700: hack/lib/cmd.sh:139: `return "${return_code}"` exited with status 1.
[INFO] 		Stack Trace: 
[INFO] 		  1: hack/lib/cmd.sh:139: `return "${return_code}"`
[INFO] 		  2: test/extended/cmd.sh:119: os::cmd::try_until_text
[INFO]   Exiting with code 1.
/home/jls/ocp-dev/src/github.com/openshift/origin/hack/lib/log/system.sh: line 31: 13997 Terminated              sar -A -o "${binary_logfile}" 1 86400 > /dev/null 2> "${stderr_logfile}"
Untagged: test/scratchimage:latest
Deleted: sha256:84d9740c38890324d9d87a1098c43bb148df9b11779fb2a9823fb460e05b3e78
[INFO] [CLEANUP] Beginning cleanup routines...
[...]
```

The thing is the oc import image done before is successful:
```
oc --config /tmp/openshift/cmd/openshift.local.config/master/admin.kubeconfig get is -n image-ns -o yaml
apiVersion: v1                                                                                                        
items:                       
- apiVersion: v1                                                                                                      
  kind: ImageStream                                        
  metadata:                                                                                                           
    annotations:             
      openshift.io/image.dockerRepositoryCheck: 2018-02-06T09:51:48Z                                                  
    creationTimestamp: 2018-02-06T09:51:48Z                
    generation: 1                                                                                                     
    name: busybox            
    namespace: image-ns                                                                                               
    resourceVersion: "1110"  
    selfLink: /oapi/v1/namespaces/image-ns/imagestreams/busybox                                                       
    uid: 5943631a-0b23-11e8-ad6d-507b9da4c680
  spec:                                                                                                               
    lookupPolicy:                                                                                                     
      local: false                                                                                                    
    tags:                    
    - annotations: null                                                                                               
      from:                         
        kind: DockerImage                                                                                             
        name: busybox:latest                                    
      generation: 1              
      importPolicy: {}                                     
      name: latest           
      referencePolicy:                                                                 
        type: Source                                       
  status:                                                  
    dockerImageRepository: 172.30.43.149:5000/image-ns/busybox
    tags:                           
    - items:
      - created: 2018-02-06T09:51:48Z
        dockerImageReference: busybox@sha256:4cee1979ba0bf7db9fc5d28fb7b798ca69ae95a47c5fecf46327720df4ff352d
        generation: 1
        image: sha256:4cee1979ba0bf7db9fc5d28fb7b798ca69ae95a47c5fecf46327720df4ff352d
      tag: latest
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

And the image is correctly tagged:

```
# docker images | grep 172.30.43.149:5000/image-ns/busybox
172.30.43.149:5000/image-ns/busybox                                      latest              5b0d59026729        13 days ago         1.15 MB
```


```# oc --config /tmp/openshift/cmd/openshift.local.config/master/admin.kubeconfig get pod -o yaml              [107/647]
apiVersion: v1                                                                                                        
items:                                                                                                                
- apiVersion: v1                                                                                                      
  kind: Pod                                               
  metadata:
    annotations:
      openshift.io/scc: restricted
    creationTimestamp: 2018-02-06T09:52:22Z
    name: new-pull-pod
    namespace: dc-ns
    resourceVersion: "1247"
    selfLink: /api/v1/namespaces/dc-ns/pods/new-pull-pod
    uid: 6d989137-0b23-11e8-ad6d-507b9da4c680
  spec:
    containers:
    - command:
      - ls
      image: 172.30.43.149:5000/image-ns/busybox
      imagePullPolicy: Always
      name: new-pull-pod
      resources: {}
      securityContext:
        capabilities:
          drop:
          - KILL
          - MKNOD
          - SETGID
          - SETUID
        privileged: false
        runAsUser: 1000100000
        seLinuxOptions:
          level: s0:c10,c5
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: default-token-h4h62
        readOnly: true
    dnsPolicy: ClusterFirst
    imagePullSecrets:
    - name: image-ns-pull
    nodeName: 192.168.32.217
    restartPolicy: Always
    schedulerName: default-scheduler
    securityContext:
      fsGroup: 1000100000
      seLinuxOptions:
        level: s0:c10,c5
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 30
    volumes:    - name: default-token-h4h62                                                                                       
      secret:                                                                                                         
        defaultMode: 420                                                                                              
        secretName: default-token-h4h62
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      message: 'containers with unready status: [new-pull-pod]'
      reason: ContainersNotReady
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      status: "True"
      type: PodScheduled
    containerStatuses:
    - image: 172.30.43.149:5000/image-ns/busybox
      imageID: ""
      lastState: {}
      name: new-pull-pod
      ready: false
      restartCount: 0
      state:
        waiting:
          message: '[rpc error: code = Unknown desc = manifest for 172.30.43.149:5000/image-ns/busybox:latest
            not found, rpc error: code = Unknown desc = manifest for 172.30.43.149:5000/image-ns/busybox:latest
            not found]'
          reason: ErrImagePull
    hostIP: 192.168.32.217
    phase: Pending
    podIP: 172.17.0.6
    qosClass: BestEffort
    startTime: 2018-02-06T09:52:22Z
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

My local docker daemon has it as insecure registry so pull shouldn't be an issue:

```
# cat /etc/sysconfig/docker | grep insecure
OPTIONS='--selinux-enabled --log-driver=journald --insecure-registry 172.30.0.0/16 --insecure-registry another.internal.registry'
```

And trying to do the pull from the daemon I get:

```
# docker pull 172.30.5.111:5000/image-ns/busybox
Using default tag: latest
Trying to pull repository 172.30.5.111:5000/image-ns/busybox ... 
manifest for 172.30.5.111:5000/image-ns/busybox:latest not found
```

BTW I'm setting SELinux to permissive while running the tests.

Any idea what can be wrong?

**Comment by juanluisvaladas** [comment]
@juanvallejo I've been delaying this far too long, I'm getting a failure while running the tests I don't understand:

```
FAILURE after 59.592s: test/extended/cmd.sh:119: executing 'oc get pods/new-pull-pod -o jsonpath='{.status.containerStatuses[0].imageID}'' expecting any result and text 'docker'; re-trying every 0.2s until completion or 60.000s: the command timed out
Standard output from the command:
Standard error from the command:
[ERROR] PID 13700: hack/lib/cmd.sh:139: `return "${return_code}"` exited with status 1.
[INFO] 		Stack Trace: 
[INFO] 		  1: hack/lib/cmd.sh:139: `return "${return_code}"`
[INFO] 		  2: test/extended/cmd.sh:119: os::cmd::try_until_text
[INFO]   Exiting with code 1.
/home/jls/ocp-dev/src/github.com/openshift/origin/hack/lib/log/system.sh: line 31: 13997 Terminated              sar -A -o "${binary_logfile}" 1 86400 > /dev/null 2> "${stderr_logfile}"
Untagged: test/scratchimage:latest
Deleted: sha256:84d9740c38890324d9d87a1098c43bb148df9b11779fb2a9823fb460e05b3e78
[INFO] [CLEANUP] Beginning cleanup routines...
[...]
```

The thing is the oc import image done before is successful:
```
oc --config /tmp/openshift/cmd/openshift.local.config/master/admin.kubeconfig get is -n image-ns -o yaml
apiVersion: v1                                                                                                        
items:                       
- apiVersion: v1                                                                                                      
  kind: ImageStream                                        
  metadata:                                                                                                           
    annotations:             
      openshift.io/image.dockerRepositoryCheck: 2018-02-06T09:51:48Z                                                  
    creationTimestamp: 2018-02-06T09:51:48Z                
    generation: 1                                                                                                     
    name: busybox            
    namespace: image-ns                                                                                               
    resourceVersion: "1110"  
    selfLink: /oapi/v1/namespaces/image-ns/imagestreams/busybox                                                       
    uid: 5943631a-0b23-11e8-ad6d-507b9da4c680
  spec:                                                                                                               
    lookupPolicy:                                                                                                     
      local: false                                                                                                    
    tags:                    
    - annotations: null                                                                                               
      from:                         
        kind: DockerImage                                                                                             
        name: busybox:latest                                    
      generation: 1              
      importPolicy: {}                                     
      name: latest           
      referencePolicy:                                                                 
        type: Source                                       
  status:                                                  
    dockerImageRepository: 172.30.43.149:5000/image-ns/busybox
    tags:                           
    - items:
      - created: 2018-02-06T09:51:48Z
        dockerImageReference: busybox@sha256:4cee1979ba0bf7db9fc5d28fb7b798ca69ae95a47c5fecf46327720df4ff352d
        generation: 1
        image: sha256:4cee1979ba0bf7db9fc5d28fb7b798ca69ae95a47c5fecf46327720df4ff352d
      tag: latest
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

And the image is correctly tagged:

```
# docker images | grep 172.30.43.149:5000/image-ns/busybox
172.30.43.149:5000/image-ns/busybox                                      latest              5b0d59026729        13 days ago         1.15 MB
```


```# oc --config /tmp/openshift/cmd/openshift.local.config/master/admin.kubeconfig get pod -o yaml              [107/647]
apiVersion: v1                                                                                                        
items:                                                                                                                
- apiVersion: v1                                                                                                      
  kind: Pod                                               
  metadata:
    annotations:
      openshift.io/scc: restricted
    creationTimestamp: 2018-02-06T09:52:22Z
    name: new-pull-pod
    namespace: dc-ns
    resourceVersion: "1247"
    selfLink: /api/v1/namespaces/dc-ns/pods/new-pull-pod
    uid: 6d989137-0b23-11e8-ad6d-507b9da4c680
  spec:
    containers:
    - command:
      - ls
      image: 172.30.43.149:5000/image-ns/busybox
      imagePullPolicy: Always
      name: new-pull-pod
      resources: {}
      securityContext:
        capabilities:
          drop:
          - KILL
          - MKNOD
          - SETGID
          - SETUID
        privileged: false
        runAsUser: 1000100000
        seLinuxOptions:
          level: s0:c10,c5
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: default-token-h4h62
        readOnly: true
    dnsPolicy: ClusterFirst
    imagePullSecrets:
    - name: image-ns-pull
    nodeName: 192.168.32.217
    restartPolicy: Always
    schedulerName: default-scheduler
    securityContext:
      fsGroup: 1000100000
      seLinuxOptions:
        level: s0:c10,c5
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 30
    volumes:    - name: default-token-h4h62                                                                                       
      secret:                                                                                                         
        defaultMode: 420                                                                                              
        secretName: default-token-h4h62
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      message: 'containers with unready status: [new-pull-pod]'
      reason: ContainersNotReady
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: 2018-02-06T09:52:22Z
      status: "True"
      type: PodScheduled
    containerStatuses:
    - image: 172.30.43.149:5000/image-ns/busybox
      imageID: ""
      lastState: {}
      name: new-pull-pod
      ready: false
      restartCount: 0
      state:
        waiting:
          message: '[rpc error: code = Unknown desc = manifest for 172.30.43.149:5000/image-ns/busybox:latest
            not found, rpc error: code = Unknown desc = manifest for 172.30.43.149:5000/image-ns/busybox:latest
            not found]'
          reason: ErrImagePull
    hostIP: 192.168.32.217
    phase: Pending
    podIP: 172.17.0.6
    qosClass: BestEffort
    startTime: 2018-02-06T09:52:22Z
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

My local docker daemon has it as insecure registry so pull shouldn't be an issue:

```
# cat /etc/sysconfig/docker | grep insecure
OPTIONS='--selinux-enabled --log-driver=journald --insecure-registry 172.30.0.0/16 --insecure-registry another.internal.registry'
```

And trying to do the pull from the daemon I get:

```
# docker pull 172.30.5.111:5000/image-ns/busybox
Using default tag: latest
Trying to pull repository 172.30.5.111:5000/image-ns/busybox ... 
manifest for 172.30.5.111:5000/image-ns/busybox:latest not found
```

BTW I'm setting SELinux to permissive while running the tests.

Any idea what can be wrong? I also have this error after rebasing against master

**Comment by juanluisvaladas** [comment]
I'm not quite sure what triggers test/extended/cli/debug.go
I'm forcing a syntax error so that I can see where this test is done in the CI

**Comment by juanluisvaladas** [comment]
test cmd

**Comment by juanvallejo** [comment]
test/cmd/debug.sh on line 27 is expecting the old error message "unable to create the debug pod ... on node ...", but instead, it's seeing the new message "error: unable to get the pod return code."

**Pull request closed without merging**